craft_archives/repo/package_repository.py

```diff
@@ -222,71 +224,81 @@ class PackageRepositoryApt(PackageRepository):
     """An APT package repository."""
 
     url: Union[AnyUrl, FileUrl]
-    key_id: KeyIdStr = pydantic.Field(alias="key-id")
-    architectures: Optional[List[str]]
-    formats: Optional[List[Literal["deb", "deb-src"]]]
-    path: Optional[str]
-    components: Optional[List[str]]
-    key_server: Optional[str] = pydantic.Field(alias="key-server")
-    suites: Optional[List[str]]
-
-    # Customize some of the validation error messages
-    class Config(PackageRepository.Config):  # noqa: D106 - no docstring needed
-        error_msg_templates = {
-            "value_error.any_str.min_length": "Invalid URL; URLs must be non-empty strings"
-        }
+    key_id: Annotated[
+        str, StringConstraints(min_length=40, max_length=40, pattern=r"^[0-9A-F]{40}$")
+    ] = Field(alias="key-id")
+    architectures: Optional[List[str]] = None
+    formats: Optional[List[Literal["deb", "deb-src"]]] = None
+    path: Optional[str] = None
+    components: Optional[List[str]] = None
+    key_server: Optional[Annotated[str, Field(alias="key-server")]] = None
+    suites: Optional[List[str]] = None
 
     @property
     def name(self) -> str:
         """Get the repository name."""
-        return re.sub(r"\W+", "_", self.url)
+        return re.sub(r"\W+", "_", str(self.url).rstrip("/"))
 
-    @validator("path")
+    @field_validator("url")
+    @classmethod
+    def _convert_url_to_string(cls, url: Union[AnyUrl, FileUrl]) -> str:
```

Comment:
what's this validator for?

Reply:
it was needed to auto strip the trailing slash of the url if any.

Reply:
but if this validator is stripping the slash, how come we need all of those other `str(url).rstrip('/')`?

Reply:
because `url` in those other occurences are of `Union[AnyUrl, FileUrl]` type, and the unit tests are failing because they are expecting a string type. As a result, I explicitly converted them to string.
I am not sure why the pydantic upgrade caused that behavior, as those tests were passing before with the union *Url type.

Reply:
I understand why the cast to string, but why the extra `strip()`? Isn't the string already stripped?
e.g. [here](https://github.com/canonical/craft-archives/pull/103/files#diff-f4b10d401baf9699bdca24e03f84d49297512fe5752a8ffc9a59d1b0711eea8dR206)

Reply:
You're right about that. I just removed them all.